File: test/Bare/TestBare.f90
program TestBare
    use Bare
    use Common
    use Fourier
    implicit none

    integer :: iorb, jorb, is, js, ik, ifreq, itau,itheta
    double precision :: omega(0:9), nu(0:9), tau1(0:9), tau2(0:9), w(4), beta , pi
    complex*16 :: fhlatt(4,4,2,5), bhlatt(4,4,2,2,5), fflatt(4,4,2,5,0:9), fflatt2(4,4,2,5,0:9), bflatt(4,4,2,2,5,0:9), bflatt2(4,4,2,2,5,0:9), &
                ftlatt(4,4,2,5,0:9), ftlatt2(4,4,2,5,0:9), btlatt(4,4,2,2,5,0:9), btlatt2(4,4,2,2,5,0:9),tempmat1(4,4), tempmat2(4,4), gfreq(0:9,4), &
                gtau(0:9,4), wfreq(0:9,4), wtau(0:9,4), fmoment(4,4,2,5,3), fhigh(4,4,2,5), bmoment(4,4,2,2,5,3), bhigh(4,4,2,2,5), err, ai

    ai = dcmplx(0.0d0, 1.0d0)
    beta=1.0d0/(8.617333262145d-5*300.0d0) 
    pi=datan2(1.0d0,1.0d0)*4.0d0 
!!!!!!!!!!!!! Fermion !!!!!!!!!!!!!!!
    fhlatt = 0.0d0
!!!!!!!!!!!!!!!!! Fermion Frequency !!!!!!!!!!!!!!!!!!!!

    omega = 0.0d0
    fflatt = 0.0d0
    fflatt2 = 0.0d0
    tempmat1 = 0.0d0
    tempmat2 = 0.0d0
    w = 0.0d0
    err = 0.0d0

    do ifreq = 0, 9                                     ! Fermion frequency condition
        omega(ifreq) = dcmplx(0.0d0, pi/beta*(2*ifreq+1))
    enddo

    do ik = 1, 5
        do is = 1, 2
            do iorb = 1, 4
                do jorb = 1, 4
                    if (iorb .eq. jorb) then
                        fhlatt(iorb, jorb, is, ik) = 1.0d0 + ik +is*0.1d0 + (iorb-jorb)*2.0d0
                    else
                        fhlatt(iorb, jorb, is, ik) = 0.1d0 + ik+ is*0.1d0 + (iorb-jorb)*0.1d0
                    endif
                enddo
            enddo
        enddo
    enddo

    do ik = 1,5
        do is = 1,2
            tempmat1 = fhlatt(:,:,is,ik)
            call hermitianeigen_dcmplx(4,w,tempmat1)

            gfreq = 0.0d0

            do iorb = 1, 4
                do ifreq = 0, 9
                    gfreq(ifreq, iorb) = 1.0d0/(ai*omega(ifreq)-w(iorb))
                enddo
            enddo

            do ifreq = 0, 9
                do iorb = 1, 4
                    do jorb = 1, 4
                        tempmat2(iorb, jorb) = tempmat1(iorb, jorb)*gfreq(ifreq, jorb)
                    enddo
                enddo
                call zgemm('n', 'c', 4, 4, 4, (1.0d0, 0.0d0), tempmat2, 4, tempmat1, 4, (0.0d0, 0.0d0), fflatt(1,1,is,ik,ifreq), 4)
            enddo
        enddo
    enddo

    call FLatFreq(4, 2, 5, 10, fhlatt,omega,fflatt2)

    print *, "FLatFreq"

    do ik = 1, 5
        do ifreq = 0, 9
            do is = 1, 2
                do iorb = 1, 4
                    do jorb = 1, 4
                        err = fflatt(iorb, jorb, is, ik, ifreq) - fflatt2(iorb, jorb, is, ik, ifreq)
                        if (cdabs(err) .gt. 1.0d-8) then
                            print '(5i5, 5f12.6)', iorb, jorb, is, ifreq, cdabs(err), fflatt(iorb, jorb, is, ik, ifreq), fflatt2(iorb, jorb, is, ik, ifreq)
                        endif
                    enddo
                enddo
            enddo
        enddo
    enddo

!!!!!!!!!!!!!!!!! Fermion Tau !!!!!!!!!!!!!!!!!!!!!!!!

    tau1 = 0.0d0
    ftlatt = 0.0d0
    ftlatt2 = 0.0d0
    tempmat1 = 0.0d0
    tempmat2 = 0.0d0
    w = 0.0d0
    err = 0.0d0
    itheta = 0.0d0
    fmoment = 0.0d0
    fhigh = 0.0d0

    do itau = 0, 9
        itheta = ttind(itau, 10)
        tau1(itau) = beta/2.0d0*(dcos(pi*(itheta+0.5d0)/dble(10))+1)
    enddo

    call FLatDyn_M(4,2,5,10,omega,fflatt,1,1,fmoment,fhigh)

    call FLatDyn_F2T(4,2,5,10,omega,fflatt,fmoment,10,tau1,ftlatt)

    call FLatTau(4,2,5,10,fhlatt,tau1,ftlatt2)

    print *, "FLatTau"

    do ik = 1, 5
        do itau = 0, 9
            do is = 1, 2
                do iorb = 1,4
                    do jorb = 1,4
                        err = ftlatt(iorb, jorb, is, ik, itau) - ftlatt2(iorb, jorb, is, ik, itau)
                        if (cdabs(err) .gt. 1.0d-8) then
                            print '(5i5, 5f12.6)', iorb, jorb, is, ik, itau, cdabs(err), ftlatt(iorb, jorb, is, ik, itau), ftlatt2(iorb, jorb, is, ik, itau)
                        endif
                    enddo
                enddo
            enddo
        enddo
    enddo

!!!!!!!!!!!!!!!!!! Boson !!!!!!!!!!!!!!!!!!!!!!!!
    bhlatt = 0.0d0
!!!!!!!!!!!!!!!!! Boson Frequency !!!!!!!!!!!!!!!!!!!!!!!!
    nu = 0.0d0
    bflatt = 0.0d0
    bflatt2 = 0.0d0
    tempmat1 = 0.0d0
    tempmat2 = 0.0d0
    w = 0.0d0
    err = 0.0d0

    do ifreq = 0,9
        nu(ifreq) = dcmplx(0.0d0, pi/beta*(2*ifreq))
    enddo

    do ik = 1, 5
        do is = 1, 2
            do js = 1, 2
                do iorb = 1, 4
                    do jorb = 1, 4
                        if (iorb .eq. jorb) then
                            bhlatt(iorb, jorb, is, js, ik) = 1.0d0 + ik + is*0.1d0 + (iorb - jorb)*2.0d0
                        else
                            bhlatt(iorb, jorb, is, js, ik) = 0.1d0 + ik + is*0.1d0 + (iorb - jorb)*0.1d0
                        endif
                    enddo
                enddo
            enddo
        enddo
    enddo

    do ik = 1, 5
        do is = 1, 2
            do js = 1, 2
                tempmat1 = bhlatt(:, :, is, js, ik)
                call hermitianeigen_dcmplx(4, w, tempmat1)

                wfreq = 0.0d0

                do iorb = 1, 4
                    do ifreq = 0, 9
                        wfreq(ifreq, iorb) = 1.0d0/(ai*nu(ifreq)-w(iorb))
                    enddo
                enddo

                do ifreq = 0, 9
                    do iorb = 1, 4
                        do jorb = 1, 4
                            tempmat2(iorb, jorb) = tempmat1(iorb, jorb)*wfreq(ifreq, jorb)
                        enddo
                    enddo
                    call zgemm('n', 'c', 4, 4, 4, (1.0d0, 0.0d0), tempmat2, 4, tempmat1, 4, (0.0d0, 0.0d0), bflatt(1,1,is,js,ik,ifreq),4)
                enddo
            enddo
        enddo
    enddo

    call BLatFreq(4,2,5,10,bhlatt,nu,bflatt2)

    print *, "BLatFreq"

    do ik = 1, 5
        do ifreq = 0, 9
            do is = 1, 2
                do js = 1, 2
                    do iorb = 1, 4
                        do jorb = 1, 4
                            err = bflatt(iorb, jorb, is, js, ik, ifreq) - bflatt2(iorb, jorb, is, js, ik, ifreq)
                            if (cdabs(err) .gt. 1.0d-8) then
                                print '(5i5, 5f12.6)', iorb, jorb, is, js, ik, ifreq, cdabs(err), bflatt(iorb, jorb, is, js, ik, ifreq), bflatt2(iorb, jorb, is, js, ik, ifreq)
                            endif
                        enddo
                    enddo
                enddo
            enddo
        enddo
    enddo

!!!!!!!!!!!!!!!!! Boson Tau !!!!!!!!!!!!!!!!!!!!!!!!
    tau2 = 0.0d0
    btlatt = 0.0d0
    btlatt2 = 0.0d0
    tempmat1 = 0.0d0
    tempmat2 = 0.0d0
    w = 0.0d0
    err = 0.0d0
    itheta = 0.0d0
    bmoment = 0.0d0
    bhigh = 0.0d0

    do itau = 0, 9
        itheta = ttind(itau, 10)
        tau2(itheta) = beta/2.0d0*(dcos(pi*(itheta+0.5d0)/dble(10))+1)
    enddo

    call BLatDyn_M(4,2,5,10,nu,bflatt,0,1,bmoment,bhigh)

    call BLatDyn_F2T(4,2,5,10,nu,bflatt,bmoment,10,tau2,btlatt)

    call BLatTau(4,2,5,10,bhlatt,tau2,btlatt2)

    print *, "BLatTau"

    do ik = 1, 5
        do itau = 0, 9
            do is = 1, 2
                do js = 1, 2
                    do iorb = 1, 4
                        do jorb = 1, 4
                            err = btlatt(iorb, jorb, is, js, ik, itau) - btlatt2(iorb, jorb, is, js, ik, itau)
                            if (cdabs(err) .gt. 1.0d-6) then 
                                print '(5i5, 5f12.6)', iorb, jorb, is, js, ik, itau, cdabs(err), btlatt(iorb, jorb, is, js, ik, itau), btlatt2(iorb, jorb, is, js, ik, itau)
                            endif
                        enddo
                    enddo
                enddo
            enddo
        enddo
    enddo
    
end program TestBare

File: class/test/TestBare.py
import numpy as np
import sys
from scipy.linalg import blas as FB 
path = "/home/momichael98/temp/Fortran/DiagE/modules"
sys.path.append(path)
import DiagE

itheta = 0.0
norb = 4
ns = 2
nk = 5
nomega = 1000
ntau = nomega

ai = 0+1j
beta = 1.0/(8.617333262145e-5*300.0)
pi = np.pi

fhlatt = np.zeros([norb,norb,ns,nk],dtype=complex,order='F')
fflatt = np.zeros([norb,norb,ns,nk,nomega],dtype=complex,order='F')
fflatt2 = np.zeros([norb,norb,ns,nk,nomega],dtype=complex,order='F')

omega = np.zeros([nomega],dtype=complex,order='F')
#gfreq = np.zeros([nomega,norb],dtype=complex,order='F')
tempmat1 = np.zeros([norb,norb],dtype=complex,order='F')
tempmat2 = np.zeros([norb,norb],dtype=complex,order='F')

err = 0.0+0.0j


for i in range(nomega):
    omega[i] = pi/beta*(2*i+1)
            
for ik in range(nk):
    for js in range(ns):
        for iorb in range(norb):
            for jorb in range(norb):
                if iorb==jorb:
                    fhlatt[iorb,jorb,js,ik] = 1.0+(ik+1)+(js+1)*0.1
                else:
                    fhlatt[iorb,jorb,js,ik] = 0.1 +(ik+1)+(js+1)*0.1 + ((iorb+1)+(jorb+1))*0.1
                                              
for ik in range(nk):
    for js in range(ns):
        tempmat1 = fhlatt[:,:,js,ik].T
        E,tempmat1 = np.linalg.eig(tempmat1) 
        gfreq = np.zeros([nomega,norb],dtype=complex,order='F') 

        for iorb in range(norb):
            for iomega in range(nomega):
                gfreq[iomega, iorb] = 1.0/(ai*omega[iomega]-E[iorb])

        for iomega in range(nomega):
            for iorb in range(norb):
                for jorb in range(norb):
                    tempmat2[iorb,jorb] = tempmat1[iorb,jorb]*gfreq[iomega,jorb]

            fflatt[:,:,js,ik,iomega] = np.matmul(tempmat2,tempmat1.T)
                 
#print(np.shape(E))
#print(fflatt)
fhlatt2 = np.array(fhlatt.T,order='F')
fflatt2 = DiagE.bare.flatfreq(fhlatt,omega)
#print(fflatt2)
print("Bare fermion frequency test")
for ik in range(nk):
    for iomega in range(nomega):
        for js in range(ns):
            for iorb in range(norb):
                for jorb in range(norb):
                    err = fflatt[iorb,jorb,js,ik,iomega]-fflatt2[iorb,jorb,js,ik,iomega]
                    if (abs(err) >= 1.0e-8):
                        print(iorb,jorb,js,ik,iomega,abs(err),fflatt[iorb,jorb,js,ik,iomega],fflatt2[iorb,jorb,js,ik,iomega])
#########################Fermion Tau####################################
tau = np.zeros([ntau],dtype=complex,order='F')
ftlatt = np.zeros([norb,norb,ns,nk,ntau],dtype=complex,order='F')
ftlatt2 = np.zeros([norb,norb,ns,nk,ntau],dtype=complex,order='F')
fmoment = np.zeros([norb,norb,ns,nk,3],dtype=complex,order='F')
fhigh = np.zeros([norb,norb,ns,nk],dtype=complex,order='F')
err = 0.0
itheta = 0.0

for itau in range(ntau):
    itheta = DiagE.common.ttind(itau,ntau)
    tau[itau] = beta/2.0*(np.cos(pi*(itheta+0.5)/ntau)+1)
fmoment, fhigh = DiagE.fourier.flatdyn_m(omega,fflatt2,1,1)

ftlatt = DiagE.fourier.flatdyn_f2t(omega,fflatt2,fmoment,tau)

ftlatt2 = DiagE.bare.flattau(fhlatt,tau)

print("Bare fermion tau test")
#print(np.shape(fmoment))
for ik in range(nk):
    for itau in range(ntau):
        for js in range(ns):
            for iorb in range(norb):
                for jorb in range(norb):
                    err = ftlatt[iorb,jorb,js,ik,itau]-ftlatt2[iorb,jorb,js,ik,itau]
                    if (abs(err) >= 1.0e-3):
                        print(iorb,jorb,js,ik,itau,abs(err),ftlatt[iorb,jorb,js,ik,itau],ftlatt2[iorb,jorb,js,ik,itau])

################ Boson Frequency ###############

nnu = nomega
nu = np.zeros([nnu],dtype=complex,order='F')
bhlatt = np.zeros([norb,norb,ns,ns,nk],dtype=complex,order='F')
bflatt = np.zeros([norb,norb,ns,ns,nk,nnu],dtype=complex,order='F')
bflatt2 = np.zeros([norb,norb,ns,ns,nk,nnu],dtype=complex,order='F')
for inu in range(nnu):
    nu[inu] = pi/beta*(2*inu)

for ik in range(nk):
    for js in range(ns):
        for ks in range(ns):
            for iorb in range(norb):
                for jorb in range(norb):
                    if iorb==jorb:
                        bhlatt[iorb,jorb,js,ks,ik] = 1.0+(ik+1)+((js+1)+(ks+1))*0.1
                    else:
                        bhlatt[iorb,jorb,js,ks,ik] = 0.1 +(ik+1)+((js+1)+(ks+1))*0.1 + ((iorb+1)+(jorb+1))*0.1
tempmat1 = np.zeros([norb,norb],dtype=complex,order='F')
tempmat2 = np.zeros([norb,norb],dtype=complex,order='F')
for ik in range(nk):
    for js in range(ns):
        for ks in range(ns):
            tempmat1 = bhlatt[:,:,js,ks,ik].T
            E, tempmat1 = np.linalg.eig(tempmat1)
            
            wfreq = np.zeros([nnu,norb],dtype=complex,order='F')

            for iorb in range(norb):
                for inu in range(nnu):
                    wfreq[inu,iorb] = 1.0/(ai*nu[inu]-E[iorb])

            for inu in range(nnu):
                for iorb in range(norb):
                    for jorb in range(norb):
                        tempmat2[iorb,jorb] = tempmat1[iorb,jorb]*wfreq[inu,jorb]
                bflatt[:,:,js,ks,ik,inu] = np.matmul(tempmat2,tempmat1.T.conj())

bflatt2 = DiagE.bare.blatfreq(bhlatt,nu)

print("Bare boson frequency")
for ik in range(nk):
    for inu in range(nnu):
        for js in range(ns):
            for ks in range(ns):
                for iorb in range(norb):
                    for jorb in range(norb):
                        err = bflatt[iorb,jorb,js,ks,ik,inu]-bflatt2[iorb,jorb,js,ks,ik,inu]
                        if abs(err) >= 1.0e-6:
                           print(iorb,jorb,js,ks,ik,inu,abs(err),bflatt[iorb,jorb,js,ks,ik,inu],bflatt2[iorb,jorb,js,ks,ik,inu])


############ Bare Boson Tau ###############
tau2 = np.zeros([ntau],dtype=complex,order='F')
btlatt = np.zeros([norb,norb,ns,ns,nk,ntau],dtype=complex,order='F')
btlatt2 = np.zeros([norb,norb,ns,ns,nk,ntau],dtype=complex,order='F')
bmoment = np.zeros([norb,norb,ns,ns,nk,3],dtype=complex,order='F')
bhigh = np.zeros([norb,norb,ns,ns,nk],dtype=complex,order='F')

for itau in range(ntau):
    itheta = DiagE.common.ttind(itau,ntau)
    tau2[itau] = beta/2.0*(np.cos(pi*(itheta+0.5)/ntau)+1)

bmoment, bhigh = DiagE.fourier.blatdyn_m(nu,bflatt2,0,1)

btlatt = DiagE.fourier.blatdyn_f2t(nu,bflatt2,bmoment,tau2)

btlatt2 = DiagE.bare.blattau(bhlatt,tau2)

print("Bare boson tau")

for ik in range(nk):
    for itau in range(ntau):
        for js in range(ns):
            for ks in range(ns):
                for iorb in range(norb):
                    for jorb in range(norb):
                        err = btlatt[iorb,jorb,js,ks,ik,itau]-btlatt2[iorb,jorb,js,ks,ik,itau]
                        if abs(err) >= 1.0e-2:
                            print(iorb,jorb,js,ks,ik,itau,abs(err),btlatt[iorb,jorb,js,ks,ik,itau],btlatt2[iorb,jorb,js,ks,ik,itau])
